src/Mod/Sketcher/Gui/Command.cpp

```diff
@@ -1213,6 +1191,250 @@ bool CmdSketcherGrid::isActive()
+            auto* sketchView = getView();
+            if (sketchView) {
+                ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/General");
+                hGrp->SetBool("SnapToGrid", state == Qt::Checked);
+            }
+        });
+
+        QObject::connect(snapAngle, qOverload<double>(&Gui::QuantitySpinBox::valueChanged), [this](double val) {
+            auto* sketchView = getView();
+
+            if (sketchView) {
+                ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/General");
+                hGrp->SetFloat("SnapAngle", val);
+            }
+        });
+
+        return snapW;
+    }
+
+private:
+    ViewProviderSketch* getView() {
+        Gui::Document* doc = Gui::Application::Instance->activeDocument();
+
+        if (doc) {
+            return dynamic_cast<SketcherGui::ViewProviderSketch*>(doc->getInEdit());
+        }
+
+        return nullptr;
+    }
+
+private:
+    QCheckBox* snapToObjects;
+    QCheckBox* snapToGrid;
+    QLabel* angleLabel;
+    Gui::QuantitySpinBox* snapAngle;
+};
+
+class CmdSketcherSnap : public Gui::Command
+{
+public:
+    CmdSketcherSnap();
+    virtual ~CmdSketcherSnap() {}
+    virtual const char* className() const
+    {
+        return "CmdSketcherSnap";
+    }
+    virtual void languageChange();
+protected:
+    virtual void activated(int iMsg);
+    virtual bool isActive(void);
+    virtual Gui::Action* createAction(void);
+private:
+    void updateIcon(bool value);
+
+    CmdSketcherSnap(const CmdSketcherSnap&) = delete;
+    CmdSketcherSnap(CmdSketcherSnap&&) = delete;
+    CmdSketcherSnap& operator= (const CmdSketcherSnap&) = delete;
+    CmdSketcherSnap& operator= (CmdSketcherSnap&&) = delete;
+};
+
+CmdSketcherSnap::CmdSketcherSnap()
+    : Command("Sketcher_Snap")
+{
+    sAppModule = "Sketcher";
+    sGroup = "Sketcher";
+    sMenuText = QT_TR_NOOP("Toggle Snap");
+    sToolTipText = QT_TR_NOOP("Toggle all snapping functionalities. In the menu you can toggle individually 'Snap to Grid', 'Snap to Objects' and further snap settings");
+    sWhatsThis = "Sketcher_Snap";
+    sStatusTip = sToolTipText;
+    eType = 0;
+}
+
+void CmdSketcherSnap::updateIcon(bool value)
+{
+    static QIcon active = Gui::BitmapFactory().iconFromTheme("Sketcher_Snap");
+    static QIcon inactive = Gui::BitmapFactory().iconFromTheme("Sketcher_Snap_Deactivated");
+
+    auto* pcAction = qobject_cast<Gui::ActionGroup*>(getAction());
+    pcAction->setIcon(value ? active : inactive);
+}
+
+void CmdSketcherSnap::activated(int iMsg)
+{
+    Q_UNUSED(iMsg);
+
+    ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/General");
+    bool value = !hGrp->GetBool("Snap", true);
+    
+    hGrp->SetBool("Snap", value);
+
+    updateIcon(value);
+
+    //Update the widget :

```

Comment:
Why is it necessary to update the widget when the button is clicked?

The widget is always updated when it is about to show. Clicking on the button when it is open closes it without action, as it is taken as cancelling the drop-down menu.

What am I missing?

Reply:
I guess it is because  you disable the controls when snap is disabled.

Reply:
Yes